src/index.ts

```diff
@@ -254,44 +301,110 @@ export default class MagmaAgent {
+                if (this.lastToolCount !== this.tools.length) {
+                    this.lastToolCount = this.tools.length;
+                    configUpdates.tools = this.tools;
+                }
 
-            // Create a new AbortController for this request
-            this.abortController = new AbortController();
+                if (this.lastSystemPromptCount !== this.fetchSystemPrompts().length) {
+                    this.lastSystemPromptCount = this.fetchSystemPrompts().length;
+                    configUpdates.system_prompts = this.fetchSystemPrompts();
+                }
 
-            const completion = await provider.makeCompletionRequest(
-                completionConfig,
-                this.onStreamChunk.bind(this),
-                0,
-                this.abortController?.signal,
-            );
+                if (Object.keys(configUpdates).length > 0) {
+                    this.sendToMagmaFlow({ type: 'config', data: configUpdates });
+                }
 
-            this.onUsageUpdate(completion.usage);
+                try {
+                    this.sendToMagmaFlow({ type: 'message', data: lastMessage });
+
+                    // Construct a promise that will resolve when a message is received from Magma Flow
+                    // Times out after kMagmaFlowMainTimeout milliseconds
+                    const serverAgentPromise = new Promise<MagmaMessage>((resolve, reject) => {
+                        // Set timeout to reject promise
+                        const timeout = setTimeout(() => {
+                            cleanup();
+                            reject(new Error('Timeout waiting for agent reply'));
+                        }, kMagmaFlowMainTimeout);
+
+                        // Handle incoming messages from Magma Flow
+                        const messageHandler = (message: MessageEvent) => {
+                            try {
+                                const data = JSON.parse(message.toString()) as SocketMessage;
+                                if (data.type === 'message' && data.data.role === 'assistant') {
+                                    cleanup();
+                                    resolve(data.data as MagmaMessage);
+                                }
+                            } catch (error) {
+                                cleanup();
+                                reject(error);
+                            }
+                        };
+
+                        // Cleanup function to clear the timeout and remove the message handler
+                        const cleanup = () => {
+                            clearTimeout(timeout);
+                            this.magmaFlowSocket?.removeListener('message', messageHandler);
+                        };
+
+                        // Add the message handler
+                        this.magmaFlowSocket.on('message', messageHandler);
+                    });
```

Comment:
Create a promise after sending a message to the server that waits for a reply from the agent. This is used when connected to magma flow rather than using the agent's local `main()` loop